server/src/main/java/com/microsoft/java/bs/core/internal/services/BuildTargetService.java

```diff
@@ -322,21 +324,39 @@ public DependencyModulesResult getBuildTargetDependencyModules(DependencyModules
     return new DependencyModulesResult(items);
   }
 
+  @FunctionalInterface
+  private interface ProgressReportCreatorType {
+    /**
+     * Create a function to create a ProgressReporter.
+     *
+     * @param targets map of Gradle TaskPaths to build targets. E.g.
+     *                `:clean` -> `plugin`, `model`, `server` or
+     *                `:model:compileJava` -> `model`
+     * @return function to create the ProgressReporter given a BuildClient.
+     */
+    Function<BuildClient, ProgressReporter> apply(Map<String, Set<BuildTargetIdentifier>> targets);
+  }
+
   /**
    * Compile the build targets.
    */
   public CompileResult compile(CompileParams params) {
     if (params.getTargets().isEmpty()) {
       return new CompileResult(StatusCode.OK);
     } else {
-      StatusCode code = runTasks(params.getTargets(), this::getBuildTaskName);
+      ProgressReportCreatorType reporterCreator =
+          fullTaskPathMap -> buildClient ->
+            new CompileProgressReporter(buildClient, params.getOriginId(), fullTaskPathMap);
```

Comment:
What do you think about:

Here we just the type of the progress reporter (like using an enum) as the argument. And then, later we can use a factory util to create the progress reporter according to the rnum type?

Reply:
I've removed `BuildClient` as a member of `GradleApiConnector` which simplifies the progress reporter creation.  It no longer uses multiple steps or passes functions.